plugin/arcconf/arcconf.py

```diff
@@ -0,0 +1,657 @@
+
+
+def _disk_type_of(arcconf_disk):
+    disk_interface = arcconf_disk['interfaceType']
+    isSSD = arcconf_disk['nonSpinning']
+
+    if isSSD is True:
+        return Disk.TYPE_SSD
+
+    if disk_interface == DISK_INTERFACE_TYPE_SATA:
+        return Disk.TYPE_SATA
+    elif disk_interface == DISK_INTERFACE_TYPE_SAS:
+        return Disk.TYPE_SAS
+
+    return Disk.TYPE_UNKNOWN
+
+
+def _disk_status_of(arcconf_disk):
+    state = arcconf_disk['state']
+
+    if state == DRIVE_READY:
+        disk_status = Disk.STATUS_OK | Disk.STATUS_FREE
+    elif state == DRIVE_ONLINE:
+        disk_status = Disk.STATUS_OK
+    else:
+        disk_status = Disk.STATUS_OTHER
+
+    return disk_status
+
+
+_ARCCONF_RAID_LEVEL_CONV = {
+    '0': Volume.RAID_TYPE_RAID0,
+    '1': Volume.RAID_TYPE_RAID1,
+    '5': Volume.RAID_TYPE_RAID5,
+    '6': Volume.RAID_TYPE_RAID6,
+    '1+0': Volume.RAID_TYPE_RAID10,
+    '50': Volume.RAID_TYPE_RAID50,
+    '60': Volume.RAID_TYPE_RAID60,
+}
+
+
+_ARCCONF_VENDOR_RAID_LEVELS = ['1adm', '1+0adm']
+
+
+_LSM_RAID_TYPE_CONV = dict(
+    list(zip(list(_ARCCONF_RAID_LEVEL_CONV.values()),
+             list(_ARCCONF_RAID_LEVEL_CONV.keys()))))
+
+
+def _arcconf_raid_level_to_lsm(arcconf_ld):
+    """
+    Based on this property:
+        Fault Tolerance: 0/1/5/6/1+0
+    """
+    arcconf_raid_level = arcconf_ld['Fault Tolerance']
+
+    if arcconf_raid_level in _ARCCONF_VENDOR_RAID_LEVELS:
+        return Volume.RAID_TYPE_OTHER
+
+    return _ARCCONF_RAID_LEVEL_CONV.get(arcconf_raid_level,
+                                        Volume.RAID_TYPE_UNKNOWN)
+
+
+def _lsm_raid_type_to_arcconf(raid_type):
+    try:
+        if raid_type == "simple_volume":
+            return Volume.RAID_TYPE_RAID0
+        else:
+            return _LSM_RAID_TYPE_CONV[raid_type]
+    except KeyError:
+        raise LsmError(
+            ErrorNumber.NO_SUPPORT,
+            "Not supported raid type %d" % raid_type)
+
+
+class Arcconf(IPlugin):
+    _DEFAULT_BIN_PATHS = [
+        "/usr/bin/arcconf",
+        "/usr/sbin/arcconf"]
+
+    def __init__(self):
+        self._arcconf_bin = None
+        self._tmo_ms = 30000
+
+    def _find_arcconf(self):
+        """
+        Try _DEFAULT_MDADM_BIN_PATHS
+        """
+        for cur_path in Arcconf._DEFAULT_BIN_PATHS:
+            if os.path.lexists(cur_path):
+                self._arcconf_bin = cur_path
+
+        if not self._arcconf_bin:
+            raise LsmError(
+                ErrorNumber.INVALID_ARGUMENT,
+                "arcconf is not installed correctly")
+
+    @_handle_errors
+    def plugin_register(self, uri, password, timeout, flags=Client.FLAG_RSVD):
+        if os.geteuid() != 0:
+            raise LsmError(
+                ErrorNumber.INVALID_ARGUMENT,
+                "This plugin requires root privilege both daemon and client")
+        uri_parsed = uri_parse(uri)
+        self._arcconf_bin = uri_parsed.get('parameters', {}).get('arcconf')
```

Comment:
Does this mean we need to try executing a command like : self._arcconf_exec(['list']) to verify _arcconf_bin actually exists?